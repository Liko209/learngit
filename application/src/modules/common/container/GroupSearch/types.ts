/*
 * @Author: Wayne Zhou (wayne.zhou@ringcentral.com)
 * @Date: 2019-08-22 16:34:05
 * Copyright Â© RingCentral. All rights reserved.
 */
import { Props } from '@/containers/Downshift/GroupSearch/GroupSearchItem/types';

export type GroupSearchProps = {
  onSelectChange: (params: { id: number }) => void;
};

export interface IGroupSearchViewModel {
  size: number;
  searchGroups: (key: string) => any;
  list: number[];
  getItemComponent: (
    id: number,
  ) => { Item: React.ComponentType<Props>; props: { itemId: number } };
}
export type GroupSearchViewProps = GroupSearchProps & IGroupSearchViewModel;
