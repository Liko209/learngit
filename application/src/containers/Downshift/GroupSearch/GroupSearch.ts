/*
 * @Author: Shining Miao (shining.miao@ringcentral.com)
 * @Date: 2018-09-18 13:46:56
 * Copyright Â© RingCentral. All rights reserved.
 */

import { buildContainer } from '@/base';
import { GroupSearchView } from './GroupSearch.View';
import { GroupSearchViewModel } from './GroupSearch.ViewModel';
import { GroupSearchProps } from './types';

const GroupSearch = buildContainer<GroupSearchProps>({
  View: GroupSearchView,
  ViewModel: GroupSearchViewModel,
});

export { GroupSearch };
