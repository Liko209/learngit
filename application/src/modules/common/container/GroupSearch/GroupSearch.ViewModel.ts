/*
 * @Author: Wayne Zhou (wayne.zhou@ringcentral.com)
 * @Date: 2019-08-22 16:34:45
 * Copyright Â© RingCentral. All rights reserved.
 */

import { GroupSearchProps, IGroupSearchViewModel } from './types';
import StoreViewModel from '@/store/ViewModel';
import { ServiceLoader, ServiceConfig } from 'sdk/module/serviceLoader';
import { SearchService } from 'sdk/module/search';
import { GroupService } from 'sdk/module/group';
import { observable, computed } from 'mobx';

export class GroupSearchViewModel extends StoreViewModel<GroupSearchProps>
  implements IGroupSearchViewModel {
  @observable list: number[] = [];
  get groupService() {
    return ServiceLoader.getInstance<GroupService>(ServiceConfig.GROUP_SERVICE);
  }

  @computed
  get size() {
    return this.list.length;
  }

  searchGroups = async (searchKey: string) => {
    const searchService = ServiceLoader.getInstance<SearchService>(
      ServiceConfig.SEARCH_SERVICE,
    );
    const result = await searchService.doFuzzySearchAllGroups(searchKey, true);
    this.list = result.sortableModels.map(item => item.id);
  };
}
