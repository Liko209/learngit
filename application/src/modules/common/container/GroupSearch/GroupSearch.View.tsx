/*
 * @Author: Wayne Zhou (wayne.zhou@ringcentral.com)
 * @Date: 2019-08-22 16:15:37
 * Copyright Â© RingCentral. All rights reserved.
 */

import React from 'react';
import { JuiGroupSearch } from 'jui/pattern/GroupSearch';
import { GroupSearchViewProps } from './types';
import portalManager from '@/common/PortalManager';
import { withAutoSizer } from 'jui/components/AutoSizer';
import { JuiVirtualizedList, IndexRange } from 'jui/components/VirtualizedList';
import { GetItemPropsOptions } from 'downshift';
import moize from 'moize';
import { UniversalAvatar } from '@/containers/Avatar/UniversalAvatar';
import { JuiEmptyScreen } from 'jui/pattern/GlobalSearch/EmptyScreen';
import { withTranslation, WithTranslation } from 'react-i18next';

const ITEM_HEIGHT = 48;

const VirtualizedListWithAutoSizer = withAutoSizer(JuiVirtualizedList);

const getItemPropsMap = {};
class GroupSearchViewComponent extends React.Component<
  GroupSearchViewProps & WithTranslation
> {
  state = {
    renderedRange: { startIndex: 0, stopIndex: 0 },
    searchKey: '',
  };

  componentDidMount() {
    this.props.searchGroups('');
  }

  private _handleChange = (e: React.ChangeEvent<any>) => {
    this.setState({ searchKey: e.target.value });
    this.props.searchGroups(e.target.value);
  };

  onSelect = (e: { id: number }) => {
    this.props.onSelect(e);
  };

  private _handleClose = () => {
    portalManager.dismissLast();
  };

  private _renderAvatar = moize(id => {
    return <UniversalAvatar id={id} />;
  });

  private _renderItem = (
    getItemProps: (options: GetItemPropsOptions<any>) => any,
    id: number,
    index: number,
    isHighlighted: boolean,
  ) => {
    getItemPropsMap[index] = getItemProps;
    if (
      this.state.renderedRange.startIndex <= index &&
      index <= this.state.renderedRange.stopIndex
    ) {
      const { Item, props } = this.props.getItemComponent(id);
      return (
        <Item
          {...props}
          key={id}
          isHighlighted={isHighlighted}
          avatar={this._renderAvatar(props.itemId)}
          data-test-automation-id="groupSearchItem"
          {...getItemProps({ index, item: { id } })}
        />
      );
    }

    return { key: id || 0 };
  };

  private _handleRenderedRangeChange = (renderedRange: IndexRange) => {
    if (
      renderedRange.startIndex !== this.state.renderedRange.startIndex ||
      renderedRange.stopIndex !== this.state.renderedRange.stopIndex
    ) {
      this.setState({ renderedRange });
    }
  };

  private _handleClear = () => {
    this.setState({ searchKey: '' });
    this.props.searchGroups('');
  };

  private _itemToString = (props: { id: number }) => {
    if (props) {
      return props.id.toString();
    }
    return '';
  };

  render() {
    const { list, t, size, dialogTitle, listTitle } = this.props;
    return (
      <JuiGroupSearch
        dialogTitle={t(dialogTitle)}
        searchPlaceHolder={t('groupSearch.inputGhostText')}
        listTitle={t(listTitle)}
        onClear={this._handleClear}
        searchKey={this.state.searchKey}
        onInputChange={this._handleChange}
        onSelectChange={this.onSelect}
        clearText={t('globalSearch.clear')}
        closeIconAriaLabel="close-icon"
        closeIconTooltip={t('common.dialog.close')}
        onDialogClose={this._handleClose}
        itemCount={this.props.size}
        onKeyDownEscape={this._handleClose}
        itemToString={this._itemToString}
      >
        {({ highlightedIndex, getItemProps }) => {
          return size === 0 ? (
            <JuiEmptyScreen
              imgWidth={60}
              textVariant="subheading1"
              text={t('globalSearch.NoMatchesFound')}
            />
          ) : (
            <VirtualizedListWithAutoSizer
              onRenderedRangeChange={this._handleRenderedRangeChange}
              fixedRowHeight={ITEM_HEIGHT}
              highlightedIndex={highlightedIndex || 0}
              overscan={2}
            >
              {list.map((id, index) =>
                this._renderItem(
                  getItemProps,
                  id,
                  index,
                  index === highlightedIndex,
                ),
              )}
            </VirtualizedListWithAutoSizer>
          );
        }}
      </JuiGroupSearch>
    );
  }
}

const GroupSearchView = withTranslation('translations')(
  GroupSearchViewComponent,
);

export { GroupSearchView };
