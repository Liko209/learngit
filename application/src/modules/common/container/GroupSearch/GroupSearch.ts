/*
 * @Author: Wayne Zhou (wayne.zhou@ringcentral.com)
 * @Date: 2019-08-22 16:20:34
 * Copyright Â© RingCentral. All rights reserved.
 */

import { buildContainer } from '@/base';
import { GroupSearchView } from './GroupSearch.View';
import { GroupSearchViewModel } from './GroupSearch.ViewModel';
import { GroupSearchProps } from './types';

const GroupSearch = buildContainer<GroupSearchProps>({
  View: GroupSearchView,
  ViewModel: GroupSearchViewModel,
});

export { GroupSearch };
