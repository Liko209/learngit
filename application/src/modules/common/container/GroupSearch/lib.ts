/*
 * @Author: Wayne Zhou (wayne.zhou@ringcentral.com)
 * @Date: 2019-08-26 23:55:59
 * Copyright Â© RingCentral. All rights reserved.
 */

import GroupModel from '@/store/models/Group';
import { CONVERSATION_TYPES } from '@/constants';
import { GroupSearchItem } from '@/containers/Downshift/GroupSearch/GroupSearchItem';
import { ContactSearchItem } from '@/containers/Downshift/ContactSearch/ContactSearchItem';
import { getGlobalValue } from '@/store/utils';
import { GLOBAL_KEYS } from '@/store/constants';
import _ from 'lodash';

export function mapGroupModelToItem(groupModel: GroupModel) {
  const members = groupModel.members;

  switch (groupModel.type) {
    case CONVERSATION_TYPES.TEAM:
    case CONVERSATION_TYPES.NORMAL_GROUP:
      return {
        Item: GroupSearchItem,
        props: { itemId: groupModel.id, size: 'large' },
      };
    case CONVERSATION_TYPES.ME:
      return {
        Item: ContactSearchItem,
        props: { itemId: members[0], showEmail: false, size: 'large' },
      };
    default: {
      const currentUserId = getGlobalValue(GLOBAL_KEYS.CURRENT_USER_ID);
      const otherId = _.difference(members, [currentUserId])[0];
      return {
        Item: ContactSearchItem,
        props: { itemId: otherId, showEmail: false, size: 'large' },
      };
    }
  }
}
