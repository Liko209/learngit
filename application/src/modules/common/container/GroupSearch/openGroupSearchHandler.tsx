/*
 * @Author: Wayne Zhou (wayne.zhou@ringcentral.com)
 * @Date: 2019-08-22 15:38:54
 * Copyright Â© RingCentral. All rights reserved.
 */

import React from 'react';
import { GroupSearch } from './GroupSearch';
import { Dialog } from '@/containers/Dialog';
import { goToConversation } from '@/common/goToConversation';
import portalManager from '@/common/PortalManager';

const DIALOG_KEY = 'GroupSearch';
function gotoConversation({ id }: { id: number }) {
  portalManager.dismissLast();
  setTimeout(() => {
    goToConversation({ conversationId: id });
  });
}

export function openGroupSearchHandler() {
  if (portalManager.isOpened(DIALOG_KEY)) return;
  Dialog.simple(
    <GroupSearch onSelectChange={gotoConversation} />,
    {
      size: 'small',
      enableEscapeClose: true,
    },
    DIALOG_KEY,
  );
  return false; // prevent browser default behavior
}
