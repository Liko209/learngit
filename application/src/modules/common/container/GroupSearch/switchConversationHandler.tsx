/*
 * @Author: Wayne Zhou (wayne.zhou@ringcentral.com)
 * @Date: 2019-08-22 15:38:54
 * Copyright Â© RingCentral. All rights reserved.
 */

import React from 'react';
import { GroupSearch } from './GroupSearch';
import { Dialog } from '@/containers/Dialog';
import { goToConversation } from '@/common/goToConversation';
import portalManager from '@/common/PortalManager';
import { analyticsCollector } from '@/AnalyticsCollector';
import { ServiceLoader, ServiceConfig } from 'sdk/module/serviceLoader';
import { SearchService } from 'sdk/module/search';

const DIALOG_KEY = 'GroupSearch';
export function switchToConversation({ id }: { id: number }) {
  portalManager.dismissAll();
  setTimeout(() => {
    analyticsCollector.goToConversation('switchConversationDialog');
    goToConversation({ conversationId: id });
  });
}

export function switchConversationHandler() {
  const searchFunc = async (searchKey: string) => {
    const searchService = ServiceLoader.getInstance<SearchService>(
      ServiceConfig.SEARCH_SERVICE,
    );
    const result = await searchService.doFuzzySearchAllGroups(searchKey, {
      myGroupsOnly: true,
      fetchAllIfSearchKeyEmpty: true,
      sortFunc: (lhs, rhs) => {
        const lhsModifiedAt = lhs.entity.most_recent_content_modified_at;
        const rhsModifiedAt = rhs.entity.most_recent_content_modified_at;

        if (lhsModifiedAt < rhsModifiedAt) return 1;
        if (lhsModifiedAt > rhsModifiedAt) return -1;
        return 0;
      },
    });
    return result.sortableModels;
  }

  analyticsCollector.shortcuts('quickSwitcher');
  if (portalManager.isOpened(DIALOG_KEY)) return;
 const {dismiss} = Dialog.simple(
    <GroupSearch onSelect={switchToConversation} dialogTitle={'groupSearch.dialogTitle'} listTitle={'groupSearch.listTitle'} searchFunc={searchFunc} />,
    {
      size: 'small',
      onClose: () => dismiss(),
      disableBackdropClick: true,
    },
    DIALOG_KEY,
  );
  return false; // prevent browser default behavior
}
