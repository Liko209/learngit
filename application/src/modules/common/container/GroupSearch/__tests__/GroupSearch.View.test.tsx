import React from 'react';
import { shallow } from 'enzyme';

import { GroupSearchView } from '../GroupSearch.View';
import toJson from 'enzyme-to-json';

describe('GroupSearch.View', () => {
  it('should render correctly [JPT-2789]', () => {
    const wrapper = shallow(
      <GroupSearchView
        size={1}
        list={[1]}
        onSelectChange={() => {}}
        searchGroups={() => {}}
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
