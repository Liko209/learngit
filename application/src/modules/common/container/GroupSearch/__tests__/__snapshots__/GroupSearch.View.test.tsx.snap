// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupSearch.View should render correctly [JPT-2789] 1`] = `
<JuiGroupSearch
  clearText="globalSearch.clear"
  closeIconAriaLabel="close-icon"
  closeIconTooltip="common.dialog.close"
  dialogTitle="groupSearch.dialogTitle"
  itemCount={1}
  itemToString={[Function]}
  listTitle="groupSearch.listTitle"
  onClear={[Function]}
  onDialogClose={[Function]}
  onInputChange={[Function]}
  onKeyDownEscape={[Function]}
  onSelectChange={[Function]}
  searchKey=""
  searchPlaceHolder="groupSearch.inputGhostText"
>
  <Component />
</JuiGroupSearch>
`;
